src/diffusers/pipelines/marigold/pipeline_marigold_depth.py

```diff
@@ -0,0 +1,784 @@
+        if self.empty_text_embedding is None:
+            self.encode_empty_text()
+
+        # 3. Preprocessing input_image
+        image, input_dtype_max = load_image_canonical(input_image)  # [N,3,H,W]
+
+        image = image.to(device=device, dtype=dtype)
+
+        original_resolution = image.shape[-2:]
+
+        if input_dtype_max is not None:
+            image = image * (2.0 / input_dtype_max) - 1.0
+        elif check_input:
+            check_image_values_range(image)
+
+        if processing_resolution > 0:
+            image = resize_to_max_edge(image, processing_resolution, resample_method_input)  # [N,3,PH,PW]
+
+        image, padding = pad_image(image, self.latent_size_scale)  # [N,3,PPH,PPW]
+
+        # 4. Encode input image into latent space. Model invocation: self.vae.encoder
+        image_latent, pred_latent = self.prepare_latent(
+            image, input_latent, generator, ensemble_size, batch_size
+        )  # [N*E,4,h,w], [N*E,4,h,w]
+
+        del image
+
+        batch_empty_text_embedding = self.empty_text_embedding.to(device=device, dtype=dtype).repeat(
+            batch_size, 1, 1
+        )  # [B,1024,2]
+
+        # 5. Denoising loop. Model invocation: self.unet
+        with self.progress_bar(total=num_images * ensemble_size * denoising_steps) as progress_bar:
+            clean_latent = []
+
+            for i in range(0, num_images * ensemble_size, batch_size):
+                batch_image_latent = image_latent[i : i + batch_size]  # [B,4,h,w]
+                batch_pred_latent = pred_latent[i : i + batch_size]  # [B,4,h,w]
+                B = batch_image_latent.shape[0]
+
+                batch_text_embedding = batch_empty_text_embedding[:B]  # [B,2,1024]
+
+                self.scheduler.set_timesteps(denoising_steps, device=device)
+
+                for t in self.scheduler.timesteps:
+                    batch_latent = torch.cat([batch_image_latent, batch_pred_latent], dim=1)  # [B,8,h,w]
+                    noise = self.unet(batch_latent, t, encoder_hidden_states=batch_text_embedding).sample  # [B,4,h,w]
+                    batch_pred_latent = self.scheduler.step(
+                        noise, t, batch_pred_latent, generator=generator
+                    ).prev_sample  # [B,4,h,w]
+                    progress_bar.update(B)
+
+                clean_latent.append(batch_pred_latent)
+
+                del batch_image_latent, batch_pred_latent, batch_text_embedding, batch_latent, noise
+
+            pred_latent = torch.cat(clean_latent, dim=0)  # [N*E,4,h,w]
+
+            del clean_latent
+
+        del image_latent, batch_empty_text_embedding
+
+        # 6. Decode prediction from latent into pixel space. Model invocation: self.vae.decoder
+        prediction = torch.cat(
+            [
+                self.decode_prediction(pred_latent[i : i + batch_size])
+                for i in range(0, pred_latent.shape[0], batch_size)
+            ],
+            dim=0,
+        )  # [N*E,1,PPH,PPW]
```

Comment:
1. No, if we want to have sequential model offloading. Currently, three stages are well-separated: (1) vae.encoder encodes everything into the latent space, (2) unet denoises everything, (3) vae.decoder decodes predictions.

Reply:
oh yeah you are right!

Reply:
2. Done: I have created `MarigoldImageProcessor` now, but I was mostly able to move the preprocessing code in it. Unfortunately, I don't see a way to move postprocessing there, because: 
 - after decoding, we unpad the predictions: this could belong to postprocessing,
 - then we ensemble - this is highly pipeline-specific and cannot be moved out,
 - Then, we format the output and visualize. As you suggested in another comment, formatting could be unified, and visualization could be completely removed.

Reply:
3: But it is not an image; it is a single-channel _prediction_ in the depth pipeline, 3-channel in the normals pipeline, and >4 channels in future pipelines. I think it makes sense to emphasize this by giving it a different name